python/ml4ir/applications/ranking/data/scripts/create_dataset.py

```diff
@@ -0,0 +1,227 @@
+    For the specified number of samples:
+    1. Generate a query key
+    2. Random sample number of results (with replacement)
+    3. Random sample to fill each result (using FeatureConfig)
+    4. TODO Check that no record ID is used twice
+    """
+
+    # Set up new DF and example data distributions for sampling options
+    rows_df = []
+    seed_dict = {}  # the keys are source data columns, the values are lists of source data column values
+    seed_dict_labeled_pos = {}  # In ranking, this is clicked vs unclicked groups
+    seed_dict_labeled_neg = {}
+
+    # Get names of essential features from feature_config
+    feature_query_key = feature_config.get_query_key('name')
+    feature_rank = feature_config.get_rank('name')
+    feature_label = feature_config.get_label('name')
+
+    # For ranking data, calculate number of results from data and filter accordingly
+    seed_data, nres = filter_nres(seed_data, feature_query_key, feature_rank, feature_nres, max_num_records, logger)
+
+    # For ranking data, calculate clicked rank from seed data as label_rank
+    seed_data = pd.merge(seed_data,
+                         seed_data[seed_data[feature_label]==True][[feature_rank,feature_query_key]].rename(columns={
+                             feature_rank:'label_rank'}),
+                         on=feature_query_key,
+                         how='left')
+
+    # Now generate data distributions from sample data
+    for mycol in list(seed_data.columns):
+        seed_dict[mycol] = list(seed_data[mycol].values)
+        seed_dict_labeled_pos[mycol] = list(seed_data[seed_data[feature_label] == 1][mycol].values)
+        seed_dict_labeled_neg[mycol] = list(seed_data[seed_data[feature_label] == 0][mycol].values)
+
+    for _ in range(num_samples):
+        # Generate a synthetic query key
+        q_id = generate_key(seed_dict, feature_config.get_query_key('name'))
+
+        # Sample to choose the number of members of the query key (in ranking, the number of results)
+        q_nseq = random.sample(seed_dict[nres], 1)[0]
+
+        # Sample to choose which member is labeled positive (in ranking, which result is clicked),
+        # sampling only from query keys with the same nseq (number of results)
+        q_labelrank = random.sample(list(seed_data[seed_data[nres]==q_nseq].label_rank.values), 1)[0]
+
+        # Build characteristics for each member of the query key (in ranking, each result of the query)
+        for i in range(int(q_nseq)):
+            is_labeled_pos = (i + 1 == q_labelrank)
+            row_dict = {
+                feature_query_key: q_id,
+                nres: q_nseq,
+                'label_rank': q_labelrank,
+                feature_rank: i + 1
+            }
+            for mycol in list(seed_data.columns):
+                # Could also let user specify distribution to sample from for each feature
+                if mycol not in row_dict.keys() and is_labeled_pos:
```

Comment:
@calliebradley What happens to features that were in the config and not in the example data? I don't think we should fill all columns in the example data, rather we should fill all columns in the feature config.features. Not sure if you are already doing that.

Reply:
@lastmansleeping Interesting, right I am replicating the features that are present in the seed data. Would you be ok making this call as a part of the L3 synthetic data story? In my mind it is peripheral to the creation of the cat failure feature. I can see that you will probably disagree with me though.

Reply:
Marking this as resolved as we agreed this will be part of the L3 synthetic data creation story. Thanks :)